### PR TITLE
support nil logger

### PR DESCRIPTION
**Pull request by navidemad**
This feature has been implemented in solid_queue gem (https://github.com/rails/solid_queue/pull/175)
I replicated the implementation for this gem.
FIxes [rails/solid_cable/issues/38](https://github.com/rails/solid_cable/issues/38)

### CODE REVIEW

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 1]
```diff
@@ -136,10 +136,18 @@ class ActionCable::SubscriptionAdapter::SolidCableTest < ActionCable::TestCase
     end
   end
 
+  test "silencing polling queries when there's no Active Record logger" do
+    with_active_record_logger(nil) do
+      @rx_adapter.send(:listener).send(:broadcast_messages)
+    end
+
+    assert true
```
Comment (navidemad) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:144]: # No logger, nothing to check, just ensure it runs without errors

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 2]
```diff
@@ -136,10 +136,18 @@ class ActionCable::SubscriptionAdapter::SolidCableTest < ActionCable::TestCase
...
+  end
+
   private
     def cable_config
       { adapter: "solid_cable", message_retention: "1.second",
-        polling_interval: "0.01.seconds" }
+        polling_interval: "0.01.seconds", silence_polling: true }
```
Comment (navidemad) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:150]: It's the default value, but needed to add it there for my mocking method with_active_record_logger

File: lib/action_cable/subscription_adapter/solid_cable.rb [hunk 3]
```diff
@@ -104,8 +104,18 @@ def broadcast_messages
             end
 
             def with_polling_volume
-              if ::SolidCable.silence_polling?
-                ActiveRecord::Base.logger.silence { yield }
+              if ::SolidCable.silence_polling? && ActiveRecord::Base.logger
```
Comment (navidemad) [inline lib/action_cable/subscription_adapter/solid_cable.rb:107]: Check if logger is not nil

File: lib/action_cable/subscription_adapter/solid_cable.rb [hunk 4]
```diff
@@ -104,8 +104,18 @@ def broadcast_messages
...
+                if ActiveRecord::Base.logger.respond_to?(:silence)
```
Comment (navidemad) [inline lib/action_cable/subscription_adapter/solid_cable.rb:108]: I was getting an error, as silence is not present in rails application (7.2.1):
```
'with_polling_volume': undefined method `silence' for an instance of Logger (NoMethodError)
ActiveRecord::Base.logger.silence { yield }
```

Comment (navidemad) [reply lib/action_cable/subscription_adapter/solid_cable.rb:108]: If this PR is merged i will do a PR to fix silence also for those who want to use solid_queue in earlier version that Rails 8

File: lib/action_cable/subscription_adapter/solid_cable.rb [hunk 5]
```diff
@@ -104,8 +104,18 @@ def broadcast_messages
             end
 
             def with_polling_volume
```
Comment (npezza93) [inline lib/action_cable/subscription_adapter/solid_cable.rb:106]: Lets have this more mirror solid queues implementation:
```ruby
          if ::SolidCable.silence_polling? && ActiveRecord::Base.logger
            ActiveRecord::Base.logger.silence { yield }
          else
            yield
          end
```

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 6]
```diff
@@ -136,10 +136,18 @@ class ActionCable::SubscriptionAdapter::SolidCableTest < ActionCable::TestCase
     end
   end
 
+  test "silencing polling queries when there's no Active Record logger" do
```
Comment (npezza93) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:139]: The assertion of this test here isnt super useful as it's just there to appease a minitest warning. Instead we should be asserting that nothing gets raised because thats what we care about. We also shouldnt be diving this deep into private methods. This test should be setup such that we change the logger to be nil, broadcast out some messages, verify the message gets broadcast and assert that nothing gets raised.

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb (hunk 2 above)
Comment (npezza93) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:150]: I dont think we need to explicitly set this here.

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 7]
```diff
@@ -159,4 +167,11 @@ def subscribe_as_queue(channel, adapter = @rx_adapter)
     ensure
       adapter.unsubscribe(channel, callback) if subscribed.set?
     end
+
+    def with_active_record_logger(logger)
+      old_logger, @server.config.logger = @server.config.logger, logger
```
Comment (npezza93) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:172]: nit but lets expand this out to make it a bit easier to read:
```suggestion
      old_logger = @server.config.logger
      @server.config.logger = logger
```

File: lib/action_cable/subscription_adapter/solid_cable.rb (hunk 5 above)
Comment (navidemad) [reply lib/action_cable/subscription_adapter/solid_cable.rb:106]: It was at the start, but when i was running it on a Rails 7.2.1 application silence was not defined inside ActiveRecord::Base

**Comment by navidemad** [comment]
@npezza93
To better reflect the intent of asserting that no exceptions are raised, I changed the test name to indicate that we are specifically testing for the absence of errors when there's no Active Record logger. Also, I ensure that the message is successfully broadcast to the queue. This approach ensures that the test covers the intended behavior without diving into private methods and focuses on preventing exceptions.

File: lib/action_cable/subscription_adapter/solid_cable.rb (hunk 5 above)
Comment (navidemad) [reply lib/action_cable/subscription_adapter/solid_cable.rb:106]: If you prefer i can do a :
`if Rails.version.to_f < 8.0`

Comment (npezza93) [reply lib/action_cable/subscription_adapter/solid_cable.rb:106]: Hmm on 7.2 this is working for me:
image

Whats your app setup like?

Comment (navidemad) [reply lib/action_cable/subscription_adapter/solid_cable.rb:106]: Yeah my mistake i was setting up in my config/environments/development.rb
`config.active_record.logger = Logger.new($stdout)`
instead of 
```
config.logger = ActiveSupport::Logger.new($stdout)
    .tap  { |logger| logger.formatter = Logger::Formatter.new }
    .then { |logger| ActiveSupport::TaggedLogging.new(logger) }
```

All good you are right i will fix the code in the PR

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 8]
```diff
@@ -136,6 +136,17 @@ class ActionCable::SubscriptionAdapter::SolidCableTest < ActionCable::TestCase
     end
   end
 
+  test "does not raise error when polling with no Active Record logger" do
+    with_active_record_logger(nil) do
+      subscribe_as_queue("channel") do |queue|
+        assert_nothing_raised do
```
Comment (npezza93) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:142]: Let's move this assertion up a level since the polling happens in between the broadcast and the pop

File: test/lib/action_cable/subscription_adapter/solid_cable_test.rb [hunk 9]
```diff
@@ -14,16 +14,16 @@ class ActionCable::SubscriptionAdapter::SolidCableTest < ActionCable::TestCase
...
       l.level = Logger::UNKNOWN
     end
 
-    adapter_klass = server.config.pubsub_adapter
+    adapter_klass = @server.config.pubsub_adapter
 
-    @rx_adapter = adapter_klass.new(server)
-    @tx_adapter = adapter_klass.new(server)
+    @rx_adapter = adapter_klass.new(@server)
+    @tx_adapter = adapter_klass.new(@server)
```
Comment (npezza93) [inline test/lib/action_cable/subscription_adapter/solid_cable_test.rb:26]: lets revert these changes since we dont need these now

**Review by npezza93 — CHANGES_REQUESTED** [review]

**Review by npezza93 — APPROVED** [review]
Thanks
